Foundation/ConsistentCasting.swift

```diff
@@ -0,0 +1,272 @@
+ |Float   |    |     |     |      |     |      |     |      |DL   |      |    |   |    |       |D       |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ |Double  |    |     |     |      |     |      |     |      |     |DL    |    |   |    |       |D       |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ |Bool    |    |     |     |      |     |      |     |      |     |      |DL  |   |    |       |D       |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ |Int     |    |     |     |      |     |      |     |      |     |      |    |DL |    |       |D       |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ |UInt    |    |     |     |      |     |      |     |      |     |      |    |   |DL  |       |D       |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ |Decimal |    |     |     |      |     |      |     |      |     |      |    |   |    |DL     |D       |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ |NSNumber|D   |D    |D    |D     |D    |D     |D    |D     |D    |D     |D   |D  |D   |       |DL      |
+ :--------+----+-----+-----+------+-----+------+-----+------+-----+------+----+---+----+-------+--------:
+ ```
+
+ - Returns: Casted representation of passed object, or `nil` if casting failed.
+
+ - Parameter object: Object to cast.
+ */
+func platformConsistentCast<T>(_ object: Any) -> T? {
+    if let casted = object as? T {
+        return casted
+    }
+
+    if let safeBridgeable = object as? _PlatformConsistentCasting {
+        return safeBridgeable._casted()
+    }
+
+    return nil
+}
+
+protocol _PlatformConsistentCasting {
+    func _casted<T>() -> T?
+}
+
+extension NSNumber: _PlatformConsistentCasting {
+    func _casted<T>() -> T? {
+        if let casted = self as? T { return casted }
+        switch T.self {
+        case is Int8.Type: return self.int8Value as? T
+        case is UInt8.Type: return self.uint8Value as? T
+        case is Int16.Type: return self.int16Value as? T
+        case is UInt16.Type: return self.uint16Value as? T
+        case is Int32.Type: return self.int32Value as? T
+        case is UInt32.Type: return self.uint32Value as? T
+        case is Int64.Type: return self.int64Value as? T
+        case is UInt64.Type: return self.uint64Value as? T
+        case is Float.Type: return self.floatValue as? T
+        case is Double.Type: return self.doubleValue as? T
+        case is Bool.Type: return self.boolValue as? T
+        case is Int.Type: return self.intValue as? T
+        case is UInt.Type: return self.uintValue as? T
+        default: return nil
+        }
+    }
+}
+
+extension Int8: _PlatformConsistentCasting {
```

Comment:
We should pull the implementation from NSNumber.swift from the overlay to make it have parity